frontends/relay-futil/compiler.py

```diff
@@ -1,90 +1,138 @@
+
+        To avoid this, we provide Relay with its own identification dictionary. If 'x' is seen
+        three times, it will produce: 'x', 'x1', x2'.
+        """
+        id_number = self.relay_id_dictionary[name]
+        self.relay_id_dictionary[name] += 1
+        if id_number == 0: return name
+        return name + str(id_number)
+
+    def produce_dahlia_name(self, name, type):
+        """
+        Dahlia uses the following naming scheme for an arbitrary variable 'X':
+        Memory1D: 'X0', 'X1', 'X2', ...
+        Memory2D: 'X0_0', 'X1_0', 'X2_0', ...
+        Memory3D: 'X0_0_0', 'X1_0_0', 'X2_0_0', ...
+        """
+        dahlia_name = self.id(name)
+        if type == PrimitiveType.Memory1D: return dahlia_name
+        if type == PrimitiveType.Memory2D: return dahlia_name + "_0"
+        if type == PrimitiveType.Memory3D: return dahlia_name + "_0_0"
+        assert False, f'{name} with {type} is not supported yet.'
+
+    def get_dahlia_declaration(self, function_name, cells, args):
+        """
+        Returns the corresponding name, Dahlia function type, and op (if it is a binary op, otherwise None).
+        If the function type isn't supported, fails with an assertion.
+        """
+        input_type = cells[0].primitive.type
+        function = name = op = None
+
+        if function_name in BuiltInBinaryCalls:
+            op = BuiltInBinaryCalls[function_name]
+            if input_type == PrimitiveType.Memory1D:
+                function = tensor1d_op
+                name = self.relay_id(f'tensor1d_{function_name}')
+            if input_type == PrimitiveType.Memory2D:
+                function = tensor2d_op
+                name = self.relay_id(f'tensor2d_{function_name}')
+        if function_name == "nn.batch_flatten":
+            if input_type == PrimitiveType.Memory3D:
+                function = tensor3d_batch_flatten
+                name = self.relay_id(f'{function.__name__}')
+        elif function_name == "nn.batch_matmul":
+            function = batch_matmul
+            name = self.relay_id(f'{function.__name__}')
+
+        assert function != None and name != None, f'{function_name} with type {input_type} is not supported.'
+        return DahliaDeclaration(component_name=name, decl_name=self.id(name), op=op, inputs=args, function=function)
 
     def visit_var(self, var):
-        name = var.name_hint
-        type = str(var.type_annotation)
-        data = [get_bitwidth(type), 1, 1]  # [width, size, index_size]
-        return [FCell(primitive=FPrimitive(name=name, data=data, type=PrimitiveType.Memory1D))]
+        name = self.relay_id(var.name_hint)
+        if self.main.contains_primitive(name): return [cell]
+
+        data, type = get_memory_parameters(var.type_annotation)
+        dahlia_name = self.produce_dahlia_name(name, type)
+        return [FCell(dahlia_name=dahlia_name, primitive=FPrimitive(name=name, data=data, type=type))]
```

Comment:
why do each of the visitor functions return a list? Can you add a comment describing this somewhere?

Reply:
This was an artifact of a different approach I took. Thanks for spotting this, I'll fix it.